source/fuzz/transformation_changing_memory_semantics.cpp

```diff
@@ -0,0 +1,194 @@
+    protobufs::TransformationChangingMemorySemantics message)
+    : message_(std::move(message)) {}
+
+TransformationChangingMemorySemantics::TransformationChangingMemorySemantics(
+    const protobufs::InstructionDescriptor& atomic_instruction,
+    uint32_t memory_semantics_operand_index,
+    uint32_t memory_semantics_new_value_id) {
+  *message_.mutable_atomic_instruction() = atomic_instruction;
+  message_.set_memory_semantics_operand_index(memory_semantics_operand_index);
+  message_.set_memory_semantics_new_value_id(memory_semantics_new_value_id);
+}
+
+bool TransformationChangingMemorySemantics::IsApplicable(
+    opt::IRContext* ir_context, const TransformationContext& /*unused*/) const {
+  auto needed_atomic_instruction =
+      FindInstruction(message_.atomic_instruction(), ir_context);
+  // The atomic instruction must exist.
+  if (!needed_atomic_instruction) {
+    return false;
+  }
+
+  // Memory semantics operand Index must be equal to 2 or 3 only.
+  auto operand_index = message_.memory_semantics_operand_index();
+  if (operand_index != 2 && operand_index != 3) {
+    return false;
+  }
+
+  // The new instruction value must exist and must be OpConstant.
+  auto new_instruction = ir_context->get_def_use_mgr()->GetDef(
+      message_.memory_semantics_new_value_id());
+  if (!new_instruction) {
+    return false;
+  }
+  if (new_instruction->opcode() != SpvOpConstant) {
+    return false;
+  }
+  auto new_memory_sematics_value = static_cast<SpvMemorySemanticsMask>(
+      new_instruction->GetSingleWordInOperand(0));
+
+  // Instruction must be atomic instruction only. Operand index must be equal to
+  // 2 in the case of instruction that takes one memory semantics operand.
+  // Instructions that takes two memory semantics, one of the value must be
+  // stronger than the other.
+  switch (needed_atomic_instruction->opcode()) {
+    case SpvOpAtomicLoad:
+    case SpvOpAtomicStore:
+    case SpvOpAtomicExchange:
+    case SpvOpAtomicIIncrement:
+    case SpvOpAtomicIDecrement:
+    case SpvOpAtomicIAdd:
+    case SpvOpAtomicISub:
+    case SpvOpAtomicSMin:
+    case SpvOpAtomicUMin:
+    case SpvOpAtomicSMax:
+    case SpvOpAtomicUMax:
+    case SpvOpAtomicAnd:
+    case SpvOpAtomicOr:
+    case SpvOpAtomicXor:
+    case SpvOpAtomicFlagTestAndSet:
+    case SpvOpAtomicFlagClear:
+    case SpvOpAtomicFAddEXT:
+      if (operand_index != 2) {
+        return false;
+      }
+      break;
+
+    case SpvOpAtomicCompareExchange:
+    case SpvOpAtomicCompareExchangeWeak:
+      if (new_memory_sematics_value ==
+          static_cast<SpvMemorySemanticsMask>(
+              ir_context->get_def_use_mgr()
+                  ->GetDef(needed_atomic_instruction->GetSingleWordInOperand(
+                      operand_index == 3 ? 2 : 3))
+                  ->GetSingleWordInOperand(0))) {
+        return false;
+      }
+      break;
+
+    default:
+      return false;
+  }
```

Comment:
I suggest that you create a static member function that checks that the opcode of |atomic_instruction| is one of the atomic instruction opcodes that we expect, and also checks that:
 - |memory_semantics_operand_index| is 0 when the opcode only supports one memory semantics;
 - |memory_semantics_operand_index| is 0 or 1 when the opcode supports two memory semantics;

I then suggest that you _assert_ these conditions in both constructors. You might think: why move the checks to the constructor and why assert these conditions? The answer is: these checks are checking the values of the protobuf (i.e. of the transformation itself). These checks do _not_ depend on the SPIR-V module (an ir_context). Thus, there is no need to do them here; we can do them at the earliest possible point (in the constructor). And, if the checks fail, then we have made a mistake in our code (e.g. in our fuzzer pass code), hence we should assert these conditions. This is very different to most IsApplicable checks, which are used to ensure that we can _try_ to apply the transformation to a slightly different SPIR-V module during shrinking, but of course it may fail, and that is OK. 

Another reason why we don't need to check the opcode is because the opcode is already checked by |FindInstruction|. Thus, we can instead assert (in the constructor) that the opcode is one of the opcodes we expect.

Reply:
Thanks for the clarification.
Done.